subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/AnnotationProcessorDetector.java

```diff
@@ -0,0 +1,128 @@
+    }
+
+    public List<AnnotationProcessorDeclaration> detectProcessors(Iterable<File> processorPath) {
+        List<AnnotationProcessorDeclaration> processors = Lists.newArrayList();
+        for (File jarOrClassesDir : processorPath) {
+            processors.addAll(cache.get(jarOrClassesDir));
+        }
+        return processors;
+    }
+
+    private static class ProcessorServiceLocator implements FileContentCacheFactory.Calculator<List<AnnotationProcessorDeclaration>> {
+        @Override
+        public List<AnnotationProcessorDeclaration> calculate(File file, FileType fileType) {
+            if (fileType == FileType.Directory) {
+                return detectProcessorsInClassesDir(file);
+            }
+            if (fileType == FileType.RegularFile && FileUtils.hasExtensionIgnoresCase(file.getName(), ".jar")) {
+                return detectProcessorsInJar(file);
+            }
+            return Collections.emptyList();
+        }
+
+        private List<AnnotationProcessorDeclaration> detectProcessorsInClassesDir(File classesDir) {
+            File processorDeclaration = new File(classesDir, "META-INF/services/javax.annotation.processing.Processor");
+            if (!processorDeclaration.isFile()) {
+                return Collections.emptyList();
+            }
+            List<String> processorNames = readLines(processorDeclaration);
+            return toProcessorDeclarations(processorNames);
+        }
+
+        private List<String> readLines(File file) {
+            try {
+                return Files.readLines(file, Charsets.UTF_8);
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+
+        private List<AnnotationProcessorDeclaration> detectProcessorsInJar(File jar) {
+            try {
+                ZipFile zipFile = new ZipFile(jar);
+                try {
+                    return detectProcessorsInZipFile(zipFile);
+                } finally {
+                    zipFile.close();
+                }
+            } catch (IOException e) {
+                DeprecationLogger.nagUserWith("Malformed jar [" + jar.getName() + "] found on compile classpath. Gradle 5.0 will no longer allow malformed jars on the compile classpath.");
+                return Collections.emptyList();
+            }
+        }
+
+        private List<AnnotationProcessorDeclaration> detectProcessorsInZipFile(ZipFile zipFile) throws IOException {
+            ZipEntry processorDeclaration = zipFile.getEntry("META-INF/services/javax.annotation.processing.Processor");
+            if (processorDeclaration == null) {
+                return Collections.emptyList();
+            }
+            List<String> processorNames = readLines(zipFile, processorDeclaration);
+            return toProcessorDeclarations(processorNames);
```

Comment:
You could also directly create the list here, by passing a `Builder` to `readLines`.

Reply:
Yes, but in the next change we'll have to correlate two different files, which would mean removing that optimization again.